Port some CoreRT Threading classes to Mono

CoffeeFlux opened the pull request:
Fixes #44795

Best reviewed commit by commit. In some cases, I've left comments in the commit description. I expect the most interesting commits to be the last few, in particular the annotations.

CoffeeFlux committed:
Remove appropriate icalls from Mono

CoffeeFlux committed:
Wire up the build

CoffeeFlux committed:
Remove old Mono managed implementations

CoffeeFlux committed:
Miscellaneous additions and changes needed for porting

These should be fairly uninteresting, so I've split them off into their own commit

CoffeeFlux committed:
Add Thread._waitInfo and switch over usage

I'm not clear on what the best practice is for internal properties like this, so I've just copied the existing style used in Thread and renamed WaitInfo to _waitInfo. Happy to change that back if WaitInfo is preferable.

CoffeeFlux committed:
Attempt to annotate the new files

I haven't had to annotate a significant amount of C# before and I'm sure I messed up somewhere, so apologies in advance. This design didn't seem very friendly to C# nullability, so it was a tough call in some places what layer to change.

vargaz commented:
In addition to the icalls, it should be possible to remove all the w32semaphore/event/mutex files from the netcore build. It might be possible to remove the whole w32handle code on platforms which have processes like wasm, i.e. w32process-unix.c is the last user of that code.

CoffeeFlux commented:
It should be yeah, but this PR is already large enough that I'd prefer to do that in a followup.

CoffeeFlux commented:
Rebased and took out of draft since the other PRs are in. Not sure if this will build, but maybe I'll get lucky?

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -198,10 +201,14 @@ public bool Join(int millisecondsTimeout)
+#if TARGET_UNIX
+            _waitInfo = new WaitSubsystem.ThreadWaitInfo(this);
+#endif
             // TODO: This can go away once the mono/mono mirror is disabled
```
I think you will need to fix this TODO. `_startHelper` can be null here given the change in the default constructor.

CoffeeFlux replied:
Ah, yep. The mirror, thankfully, is now disabled :)

CoffeeFlux committed:
Fix todo

Not sure if I can get rid of the property altogether? Waiting on aleksey to get back to me about the threading changes

vargaz commented on src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.ThreadWaitInfo.Unix.cs:
```diff
@@ -395,7 +395,7 @@ public static void Sleep(int timeoutMilliseconds, bool interruptible)
                 if (timeoutMilliseconds == 0)
                 {
-                    if (interruptible && Thread.CurrentThread.WaitInfo.CheckAndResetPendingInterrupt_NotLocked)
+                    if (interruptible && Thread.CurrentThread._waitInfo.CheckAndResetPendingInterrupt_NotLocked)
```
Having an internal WaitInfo property is fine imho.

CoffeeFlux replied:
I wasn't sure which convention was preferred in dotnet/runtime; it's easy to change in either direction.

CoffeeFlux replied:
@marek-safar do you know which convention is preferred?

stephentoub replied:
Either is ok.  Properties are a tad more expensive for the JIT as they need to be inlined to get the right perf.  But especially between components it's often worth it for the flexibility of being able to easily change the property's implementation without having to also find and update all call sites.

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -198,12 +201,14 @@ public bool Join(int millisecondsTimeout)
             InitInternal(this);
 
-            // TODO: This can go away once the mono/mono mirror is disabled
-            stack_size = _startHelper!._maxStackSize;
```
I think you also need to delete the `stack_size` field and pass the `stack_size` as argument instead.

CoffeeFlux replied:
I pinged Aleksey about this TODO since I don't have context for the recent Thread changes and don't want to spent half an hour looking through it all if possible. I'll update tomorrow after I talk with him.

vargaz replied:
This is currently used by the runtime.

lambdageek replied:
> I pinged Aleksey about this TODO since I don't have context for the recent Thread changes and don't want to spent half an hour looking through it all if possible. I'll update tomorrow after I talk with him.

I don't remember what that comment means anymore except the superficial "once we stop the mirror, the layout of `MonoInternalThread` can change".

But yea we so still need to pass a stack_size down into the runtime so that it's threaded down to the native thread creation functions. (otherwise it ends up initialized to 0 which I think mean "default size")

jkotas replied:
I have added this comment one month ago as I was moving more Thread.cs to be shared. It is exactly as @lambdageek says - you need to pass a stack_size down into the native thread creating function.

CoffeeFlux replied:
Hopefully I understood correctly how the stack size stuff is supposed to work?

CoffeeFlux committed:
Fix Windows build

Another ifdef is needed

vargaz commented:
The runtime changes look ok to me.

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.cs:
```diff
@@ -409,6 +409,13 @@ private static bool SignalAndWait(WaitHandle toSignal, WaitHandle toWaitOn, int
             }
         }
 
+        internal static void ThrowInvalidHandleException()
```
```csharp
internal static void ThrowInvalidHandleException() => throw new InvalidOperationException(SR.InvalidOperation_InvalidHandle) { HResult = HResults.E_HANDLE; }
```

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Threading/EventWaitHandle.Unix.cs:
```diff
@@ -19,7 +19,7 @@ private void CreateEventCore(bool initialState, EventResetMode mode, string name
         }
 
-        private static OpenExistingResult OpenExistingWorker(string name, out EventWaitHandle result)
+        private static OpenExistingResult OpenExistingWorker(string name, out EventWaitHandle? result)
```
This might be better to move to the caller. It seems to do always do some error checking which is Windows specific

CoffeeFlux replied:
I think we have to match the Windows equivalent here, since otherwise you end up with a bunch of weird platform-specific code at a higher level? Maybe I'm misunderstanding.

marek-safar replied:
I was referring to code patterns like https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/System.Private.CoreLib/src/System/Threading/EventWaitHandle.cs#L32-L45. The API is windows only and on Unix will have never reached error handling.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs:
```diff
@@ -220,8 +220,8 @@ private void AcceptSignal(ThreadWaitInfo waitInfo)
-                        Debug.Assert(_ownershipInfo.Thread == null);
-                        _ownershipInfo.AssignOwnership(this, waitInfo);
+                        Debug.Assert(_ownershipInfo!.Thread == null);
+                        _ownershipInfo!.AssignOwnership(this, waitInfo);
```
I'm surprised this `!` is required; I'd have thought the one on the previous line would have negated the need for this one.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs:
```diff
@@ -744,9 +744,9 @@ private void SignalMutex(bool isAbandoned)
                 Debug.Assert(!IsSignaled);
 
-                _ownershipInfo.RelinquishOwnership(this, isAbandoned);
+                _ownershipInfo!.RelinquishOwnership(this, isAbandoned);
```
Rather than adding these `!`s, can you add a `Debug.Assert(_ownershipInfo != null);` above?

stephentoub commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLock.cs:
```diff
@@ -36,5 +36,21 @@ public void VerifyIsLocked()
+        }
+
+#if DEBUG
+        public bool IsLocked
```
Where is this used?

CoffeeFlux replied:
In `src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.ThreadWaitInfo.Unix.cs`:
`Debug.Assert(s_lock.IsLocked || _waitMonitor.IsLocked);`

CoffeeFlux replied:
I only added this after the initial PR when I noticed build errors on CI, since locally I was just testing release (and so this didn't get hit).

CoffeeFlux commented:
Oh yeah, cc: @fanyang-mono @naricc, as per usual I'm changing threading things and this might have perf implications, either for TE or microbenchmarks. No need to run in advance of merging this PR, but you might see movement.

CoffeeFlux committed:
Back to WaitInfo

CoffeeFlux committed:
Annotation fixes

CoffeeFlux committed:
Add WaitHandle.Windows.cs

This is what I get for not testing on a Windows machine, can split into a separate PR if needed

CoffeeFlux committed:
Rename to just FeatureManagedThreading and add WaitHandle.Windows.cs

CoffeeFlux committed:
Thread down stack size as part of thread creation

CoffeeFlux commented:
Ignore the Windows stuff for now, I'll just boot a Windows machine and sort out whatever build issues are arising there.

jkotas commented on src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -110,6 +110,7 @@
     <FeaturePortableTimer Condition="'$(TargetsBrowser)' != 'true'">true</FeaturePortableTimer>
     <FeaturePortableThreadPool Condition="'$(TargetsBrowser)' != 'true'">true</FeaturePortableThreadPool>
     <FeaturePerfTracing Condition="'$(TargetsBrowser)' != 'true'">true</FeaturePerfTracing>
+    <FeatureManagedThreading>true</FeatureManagedThreading>
```
I do not think this deserves own Feature flag. The two places where it is used can just use `'$(FeatureCoreCLR)' != 'true'`. It won't be the first condition like that in System.Private.CoreLib.Shared.projitems .

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -67,6 +66,9 @@ public partial class Thread
         internal ExecutionContext? _executionContext;
         internal SynchronizationContext? _synchronizationContext;
+#if TARGET_UNIX
+        internal WaitSubsystem.ThreadWaitInfo WaitInfo;
```
Nit: To follow the coding convention, this should be internal WaitSubsystem.ThreadWaitInfo WaitInfo => _waitInfo;

CoffeeFlux committed:
No separate feature flag

CoffeeFlux committed:
Expose property correctly

CoffeeFlux committed:
Add Windows Interop files

CoffeeFlux committed:
Disable reentrant wait functionality

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs:
```diff
@@ -43,7 +45,6 @@ private static unsafe int WaitForMultipleObjectsIgnoringSyncContext(IntPtr* pHan
             }
 
-            Thread currentThread = Thread.CurrentThread;
             currentThread.SetWaitSleepJoinState();
```
I think you want to keep SetWaitSleepJoinState outside the MONO ifdef.

jkotas commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.Threading.cs:
```diff
@@ -0,0 +1,75 @@
+        internal const int WAIT_FAILED = unchecked((int)0xFFFFFFFF);
+
+        [DllImport(Libraries.Kernel32)]
+        internal static extern uint WaitForMultipleObjectsEx(uint nCount, IntPtr lpHandles, bool bWaitAll, uint dwMilliseconds, bool bAlertable);
```
It looks like that this is trying to minimize marshaling overhead by calling `GetLastError` directly. It may make sense to also change the `bool` arguments to use `BOOL` to eliminate the rest of the marshaling.

CoffeeFlux committed:
Attempt to fix wasm

CoffeeFlux committed:
Fix ifdefs

CoffeeFlux committed:
Switch bool to BOOL for interop

CoffeeFlux committed:
Always create _waitInfo

CoffeeFlux commented:
Seeing some failures like this:
```
    System.IO.IsolatedStorage.IsolatedStorageFileStreamTests.DisposeAsync_MultipleInvokes_Idempotent(scope: UserStoreForApplication) [FAIL]
      System.TypeInitializationException : The type initializer for 'System.IO.IsolatedStorage.TestHelper' threw an exception.
      ---- System.PlatformNotSupportedException : The named version of this synchronization primitive is not supported on this platform.
      Stack Trace:
        /_/src/libraries/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/IsolatedStorageFileStreamTests.cs(14,0): at System.IO.IsolatedStorage.IsolatedStorageFileStreamTests.DisposeAsync_MultipleInvokes_Idempotent(PresetScopes scope)
        --- End of stack trace from previous location ---
        ----- Inner Stack Trace -----
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/Mutex.Unix.cs(22,0): at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/Mutex.cs(23,0): at System.Threading.Mutex..ctor(Boolean initiallyOwned, String name)
        /_/src/libraries/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32Unix.cs(135,0): at System.IO.IsolatedStorage.Helper.CreateMutexNotOwned(String pathName)
        /_/src/libraries/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/Helper.Win32Unix.cs(78,0): at System.IO.IsolatedStorage.Helper.GetRandomDirectory(String rootDirectory, IsolatedStorageScope scope)
        /_/src/libraries/System.IO.IsolatedStorage/tests/System/IO/IsolatedStorage/TestHelper.cs(27,0): at System.IO.IsolatedStorage.TestHelper..cctor()
```

I don't believe I touched that behavior, so this is just a check that did not exist in the Mono version. The helper being used does create a named mutex, and looking at the Windows-specific behavior it involves I'm surprised this test doesn't break elsewhere.

Is the check here from CoreRT intended to be the behavior globally, or should I just remove that check and silently ignore the name? If it's the intended behavior, we probably need to update the docs.

CoffeeFlux commented:
There seem to be two other interesting classes of failure.

SemaphoreFullException:
```
    System.Threading.Tests.WaitHandleTests.SignalAndWait(toSignal: Semaphore { Handle = 3958, SafeWaitHandle = SafeWaitHandle { IsClosed = False, IsInvalid = False } }, toWaitOn: AutoResetEvent { Handle = 3974, SafeWaitHandle = SafeWaitHandle { IsClosed = False, IsInvalid = False } }, callSignalAndWait: Func`3 { Method = Boolean <SignalAndWait_MemberData>b__24_0(System.Threading.WaitHandle, System.Threading.WaitHandle), Target = <>c { } }, skipInfiniteWaitTests: True) [FAIL]
      Assert.Throws() Failure
      Expected: typeof(System.InvalidOperationException)
      Actual:   typeof(System.Threading.SemaphoreFullException): Adding the specified count to the semaphore would cause it to exceed its maximum count.
      ---- System.Threading.SemaphoreFullException : Adding the specified count to the semaphore would cause it to exceed its maximum count.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs(701,0): at System.Threading.WaitSubsystem.WaitableObject.SignalSemaphore(Int32 count)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs(569,0): at System.Threading.WaitSubsystem.WaitableObject.Signal(Int32 count)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs(378,0): at System.Threading.WaitSubsystem.SignalAndWait(WaitableObject waitableObjectToSignal, WaitableObject waitableObjectToWaitOn, Int32 timeoutMilliseconds, Boolean interruptible, Boolean prioritize)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs(349,0): at System.Threading.WaitSubsystem.SignalAndWait(IntPtr handleToSignal, IntPtr handleToWaitOn, Int32 timeoutMilliseconds)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Unix.cs(17,0): at System.Threading.WaitHandle.SignalAndWaitCore(IntPtr handleToSignal, IntPtr handleToWaitOn, Int32 millisecondsTimeout)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.cs(387,0): at System.Threading.WaitHandle.SignalAndWait(WaitHandle toSignal, WaitHandle toWaitOn, Int32 millisecondsTimeout)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.cs(451,0): at System.Threading.WaitHandle.SignalAndWait(WaitHandle toSignal, WaitHandle toWaitOn)
        /_/src/libraries/System.Runtime/tests/System/Threading/WaitHandleTests.cs(304,0): at System.Threading.Tests.WaitHandleTests.<>c.<SignalAndWait_MemberData>b__24_0(WaitHandle s, WaitHandle w)
        /_/src/libraries/System.Runtime/tests/System/Threading/WaitHandleTests.cs(367,0): at System.Threading.Tests.WaitHandleTests.<>c__DisplayClass25_0.<SignalAndWait>b__1()
        ----- Inner Stack Trace -----
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs(701,0): at System.Threading.WaitSubsystem.WaitableObject.SignalSemaphore(Int32 count)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs(569,0): at System.Threading.WaitSubsystem.WaitableObject.Signal(Int32 count)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs(378,0): at System.Threading.WaitSubsystem.SignalAndWait(WaitableObject waitableObjectToSignal, WaitableObject waitableObjectToWaitOn, Int32 timeoutMilliseconds, Boolean interruptible, Boolean prioritize)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs(349,0): at System.Threading.WaitSubsystem.SignalAndWait(IntPtr handleToSignal, IntPtr handleToWaitOn, Int32 timeoutMilliseconds)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Unix.cs(17,0): at System.Threading.WaitHandle.SignalAndWaitCore(IntPtr handleToSignal, IntPtr handleToWaitOn, Int32 millisecondsTimeout)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.cs(387,0): at System.Threading.WaitHandle.SignalAndWait(WaitHandle toSignal, WaitHandle toWaitOn, Int32 millisecondsTimeout)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.cs(451,0): at System.Threading.WaitHandle.SignalAndWait(WaitHandle toSignal, WaitHandle toWaitOn)
        /_/src/libraries/System.Runtime/tests/System/Threading/WaitHandleTests.cs(304,0): at System.Threading.Tests.WaitHandleTests.<>c.<SignalAndWait_MemberData>b__24_0(WaitHandle s, WaitHandle w)
        /_/src/libraries/System.Runtime/tests/System/Threading/WaitHandleTests.cs(367,0): at System.Threading.Tests.WaitHandleTests.<>c__DisplayClass25_0.<SignalAndWait>b__1()
```

No AbandonedMutexException:
```
    System.Threading.Tests.MutexTests.AbandonExisting(name: null, waitType: WaitOne, waitCount: 1, notAbandonedWaitIndex: 0, isNotAbandonedWaitObjectSignaled: False, abandonDuringWait: False) [FAIL]
      System.AggregateException : One or more errors occurred. (Result: False (Assert.Throws() Failure
      Expected: typeof(System.Threading.AbandonedMutexException)
      Actual:   (No exception was thrown)))
      ---- System.AggregateException : Result: False (Assert.Throws() Failure
      Expected: typeof(System.Threading.AbandonedMutexException)
      Actual:   (No exception was thrown))
      -------- Assert.Throws() Failure
      Expected: typeof(System.Threading.AbandonedMutexException)
      Actual:   (No exception was thrown)
      Stack Trace:
        /_/src/libraries/Common/tests/System/Threading/ThreadTestHelpers.cs(51,0): at System.Threading.Tests.ThreadTestHelpers.<>c__DisplayClass4_0.<CreateGuardedThread>b__1()
        /_/src/libraries/Common/tests/System/Threading/ThreadTestHelpers.cs(56,0): at System.Threading.Tests.ThreadTestHelpers.<>c__DisplayClass4_0.<CreateGuardedThread>b__2()
        /_/src/libraries/Common/tests/System/Threading/ThreadTestHelpers.cs(107,0): at System.Threading.Tests.ThreadTestHelpers.RunTestInBackgroundThread(Action test)
        /_/src/libraries/System.Threading/tests/MutexTests.cs(237,0): at System.Threading.Tests.MutexTests.AbandonExisting(String name, WaitHandleWaitType waitType, Int32 waitCount, Int32 notAbandonedWaitIndex, Boolean isNotAbandonedWaitObjectSignaled, Boolean abandonDuringWait)
        /_/src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(378,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
        ----- Inner Stack Trace -----
        /_/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs(114,0): at System.AssertExtensions.Throws[AbandonedMutexException,Boolean](Func`1 func)
        /_/src/libraries/System.Threading/tests/MutexTests.cs(295,0): at System.Threading.Tests.MutexTests.<>c__DisplayClass13_0.<AbandonExisting>b__0()
        /_/src/libraries/Common/tests/System/Threading/ThreadTestHelpers.cs(35,0): at System.Threading.Tests.ThreadTestHelpers.<>c__DisplayClass4_0.<CreateGuardedThread>b__0()
        ----- Inner Stack Trace -----
        /_/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs(91,0): at System.AssertExtensions.Throws[AbandonedMutexException,Boolean](Func`1 func)
```

jkotas commented:
> Is the check here from CoreRT intended to be the behavior globally, or should I just remove that check and silently ignore the name? If it's the intended behavior, we probably need to update the docs.

Right, the named semaphores are not supported by CoreRT on non-Windows. I believe that Mono did not support them either, but it was silently ignoring the name instead of throwing an exception. I think that the correct behavior for Mono is to throw the exception.

CoffeeFlux commented:
I guess I need to disable these tests on Mono then, and if CoreCLR is planning to migrate to the CoreRT implementation as well I'll open an issue to fix the tests.

jkotas commented:
> CoreCLR is planning to migrate to the CoreRT implementation as well I'll open an issue to fix the tests.

I think we would have to implement the support for named mutexes in the managed impl before CoreCLR can migrate to it.

CoffeeFlux commented:
Gotcha. I'll just make an issue to add support for that then.

CoffeeFlux commented:
There are also additional tests in here that can probably be re-enabled, but I'll do that in a followup PR.

CoffeeFlux committed:
Some exclusions

CoffeeFlux committed:
Re-enable the tests and disable the name check

Some of these are uncovering real issues, so I think it's better to silently ignore the argument for now

CoffeeFlux committed:
Remove thread_interrupt_requested

This might be a bad idea and we should just call into managed to check?

CoffeeFlux committed:
Correctly abandon mutexes

This is probably not optimal from a linker perspective, but that can be dealt with later

CoffeeFlux committed:
Disable tests again

The remaining failures are just related to the named mutexes

CoffeeFlux commented:
While both cases now pass, these are the minimal examples I've been using for 2/3 failures: https://gist.github.com/CoffeeFlux/46abf0c447ba19a249d5fde217f8b802

I can't reproduce the `SemaphoreFullException` locally on my Mac for whatever reason, and have no clue what would cause that. I'll try Linux tomorrow I guess...

CoffeeFlux committed:
Uncomment unsupported name exception

CoffeeFlux committed:
Excise a lot more runtime code

Yeehaw

CoffeeFlux committed:
Disable another test

CoffeeFlux committed:
Windows build fix

CoffeeFlux commented:
OK, remaining failures:

SemaphoreFullException (linux-only): https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-47333-merge-d26a4046cb174e31a9/System.Runtime.Tests/console.f59fcfa3.log?sv=2019-07-07&se=2021-03-18T06%3A50%3A40Z&sr=c&sp=rl&sig=eMTD0rEu3ILAkMynDmI92VJCiAsZ6hFejig7aGmCZkM%3D

ThreadInterruptException (windows-only): https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-47333-merge-a8ed1943579d463ab3/System.Threading.Thread.Tests/console.6b801d7b.log?sv=2019-07-07&se=2021-03-18T06%3A50%3A57Z&sr=c&sp=rl&sig=5gO6PNSEyv1o9umXleFkE1%2BU9AaZYrTunMmpS8lHtu0%3D

abandon_mutexes crash (probably just needs an extra guard): https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-47333-merge-d481cc0666874a0ea9/Interop/console.0775828b.log?sv=2019-07-07&se=2021-03-18T07%3A50%3A15Z&sr=c&sp=rl&sig=H%2Fs5wKo5kntVvhjGv1rBBCgLseU4bxvtKvZAUtpmwVg%3D

Inscrutable Linux-only crashes in networking tests(???) that do seem related: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-47333-merge-36eb2a7af9234001b3/System.Net.NameResolution.Functional.Tests/console.869f5807.log?sv=2019-07-07&se=2021-03-18T07%3A01%3A00Z&sr=c&sp=rl&sig=oBuA99OI6dYKZFh9oCl3xBvC0me%2BnMlchZw%2FehvPoFY%3D

lambdageek commented on src/mono/mono/metadata/threads.c:
```diff
@@ -1309,8 +1308,6 @@ create_thread (MonoThread *thread, MonoInternalThread *internal, MonoThreadStart
 		goto done;
 	}
 
-	internal->stack_size = (int) stack_set_size;
```
nit: `mono_thread_platform_create_thread` for a bunch of platforms has code or comments that modifies the stack size argument to return the actual stack size of the new thread.  If we're not using that info anywhere anymore we could simplify `mono_thread_platform_create_thread`

(I assume we still have a way to query the stack size through other means - I think the Task subsystem needs to look it up, for example - instead of looking at the stack_size field on the thread object)

lambdageek commented on src/mono/mono/metadata/threads.c:
```diff
@@ -4096,15 +4061,6 @@ mono_thread_execute_interruption (MonoExceptionHandle *pexc)
 		/* calls UNLOCK_THREAD (thread) */
 		self_suspend_internal ();
 		unlock = FALSE;
-	} else if (MONO_HANDLE_GETVAL (thread, thread_interrupt_requested)) {
```
I think when you get rid of this, you make `mono_sleep_internal` uninterruptible - it calls this function inside its sleep loop.

lambdageek commented on src/mono/mono/metadata/w32mutex-unix.c:
```diff
@@ -253,180 +250,41 @@ mono_w32mutex_init (void)
-			__func__, mono_w32handle_get_typename (handle_data->type), handle);
+	gpointer args [1];
+	args [0] = internal;
+	mono_runtime_try_invoke_handle (thread_exiting, NULL_HANDLE, args, error);
```
`mono_w32mutex_abandon` is called from `mono_thread_detach_internal` which is called from, for example, the TLS key dtor for the MonoThreadInfo (via the `thread_detach` callback). If that happens, it's might be too late to run managed code.

vargaz replied:
On corert, the runtime calls the OnThreadExit () callback in Thread.CoreRT.Unix.cs when the thread exists. This is called on the thread itself just before it exits.

lambdageek replied:
yea for managed threads I agree it makes sense to just call it from managed.

but for a foreign thread that just called some reverse pinvoke and doesn't know anything about Mono, if it exits while holding a mutex, our only chance to cleanup is through the TLS destructor.  I'd be curious to know what CoreRT does in that case.

jkotas replied:
CoreRT just calls into managed code in this case. We can reason about the bit of managed code that runs in this case. We know that it has been fully AOT and there will be no JITing or other heavy weight operation.

vargaz replied:
What we do in a lot of cases in mono is do the operation from the finalizer thread, i.e. see mono_gc_finalize_notify ().

lambdageek commented on src/mono/mono/metadata/threads.c:
```diff
@@ -869,6 +868,42 @@ mono_thread_attach_internal (MonoThread *thread, gboolean force_attach, gboolean
+
+	gpointer args [1];
+	args [0] = internal;
+	mono_runtime_try_invoke_handle (thread_exiting, NULL_HANDLE, args, error);
```
So we saw a crash here in `Running TestUnmanagedCallersOnlyValid_OnNewNativeThread...` which is a good test - it runs some managed code on a foreign native thread.  Which means that the runtime will still have a `MonoThreadInfo` with a TLS dtor on that foreign thread which will call `mono_thread_detach_internal` (via the `thread_detach` callback) which will come here.

But by the time we get here, the domain is already unset.  So the JIT will die.

```
=================================================================
      Native stacktrace:
      =================================================================
      0x10ee5e6d6 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_dump_native_crash_info
      0x10ee024ae - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_handle_native_crash
      0x10ed5a515 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_sigsegv_signal_handler_debug
      0x7fff69677f5a - /usr/lib/system/libsystem_platform.dylib : _sigtramp
      0x0 - Unknown
      0x10ed59fb1 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : lookup_method
      0x10ed58c92 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_jit_compile_method_with_opt
      0x10ed5d09f - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_jit_runtime_invoke
      0x10ec6215f - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_runtime_try_invoke
      0x10ec643fe - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_runtime_try_invoke_handle
      0x10ec7825b - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : abandon_mutexes
      0x10ec7513c - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_thread_detach_internal
      0x10ec77cdd - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : thread_detach
      0x10ece23c4 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : unregister_thread
      0x10ece2c20 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : thread_info_key_dtor
      0x7fff69683163 - /usr/lib/system/libsystem_pthread.dylib : _pthread_tsd_cleanup
      0x7fff69682ee9 - /usr/lib/system/libsystem_pthread.dylib : _pthread_exit
      0x7fff6968166c - /usr/lib/system/libsystem_pthread.dylib : pthread_sigmask
      0x7fff6968150d - /usr/lib/system/libsystem_pthread.dylib : _pthread_body
      0x7fff69680bf9 - /usr/lib/system/libsystem_pthread.dylib : thread_start
 ```

---

So the gist of the current approach is: "call OnThreadExiting" from the dying thread.

I think that's not a workable idea - we already started tearing down the thread and calling managed code will be fragile at this point.

Instead I think we should do this: create a finalizable managed object that is immediately garbage and give it the responsibility to call OnThreadExiting by giving it a reference to the managed thread object.

We have the managed thread object still, because (except in the case of shutdown) it's not destroyed when the native thread is destroyed - it goes away when the last reference to it goes away - but the `MonoThreadInfo` is still holding a handle to it.

The new object will be created and it won't have any references to it as soon as the native thread exits. So on the next GC it will be garbage.  When the finalizer runs on the finalizer thread, it should abandon the mutexes.  (I guess one complication is that we can't just put a managed reference to the thread into the object since it'll be null when the finalizer runs - so maybe put a GC Handle in there - that seems dangerous, but maybe it's ok?)

lambdageek replied:
I guess we can't actually guarantee there will be a GC and that the finalizers will run

CoffeeFlux committed:
Disable InterruptTest on Windows

CoffeeFlux committed:
Add link to issue on reentrant wait support for Windows

CoffeeFlux committed:
Move mutex abandoning to the managed Thread object's finalizer

I originally implemented Zoltan's suggestion, but this seems to work and is significantly cleaner. Pre-jitting is also an option here, so if this doesn't work try that next.

CoffeeFlux committed:
Rethrow with InvalidOperationException when called from WaitHandle

I think it would make much more sense to always throw SemaphoreFullException instead of sometimes throwing InvalidOperationException saying the semaphore is full, but this preserves the old behavior.

CoffeeFlux committed:
Another ActiveIssue on Windows

CoffeeFlux committed:
Linker exclusion no longer needed

CoffeeFlux commented:
I still have a cleanup-related change to push, but this is green on CI. Issues uncovered as part of the port:

https://github.com/dotnet/runtime/issues/49518

https://github.com/dotnet/runtime/issues/49521

https://github.com/dotnet/runtime/issues/48720

I don't consider any of these a big enough deal to block a merge here. All of them would probably block CoreCLR from adopting the managed implementation, but from a Mono perspective the only regression is the interrupts on Windows, and thankfully that appears to be the easiest of the three.

For the mutex abandonment, I opted to do it from the Thread object's destructor. As far as I can tell this works fine and is fairly clean, but if the approach is discovered to be faulty I can switch to pre-jitting the method instead.

For the different exception types, I just catch and rethrow. I'm not particularly fond of throwing an `InvalidOperationException` saying the semaphore is full instead of just a `SemaphoreFullException`, but that's the legacy behavior.

jkotas reviewed (approved):
Nice!
